### PR TITLE
MM 33398 - Implement Backup supervisor

### PR DESCRIPTION
**Pull request by Szymongib**
#### Summary

This PR introduces Backup Supervisor, which performs backup operations for installations.
PR bullet points:
- Backup is run as Kubernetes job deployed on Cluster on which one of the Cluster Installations run, in this Installation's namespace.
- The Installation needs to be in `hibernating` state to trigger backup (will also be checked on API level).
- After backup is triggered `InstallationBackup.ClusterInstallationID` is set to the id  of Cluster Installation next to which it is running.
- The Installation state can change after backup started running.
- Backup is marked as `failed` if the job fails and as `succeeded` if it completed.

Besides locking mechanism in this PR, to ensure that backup does not end in unpredictable state upcoming PRs will introduce following constraints:
- Cluster Installation cannot be deleted if Backup is running "on it".
- Installation cannot be deleted if Backup is requested or in progress for it.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-33398

#### Release Note

```release-note
Introduce Backup supervisor
```

### CODE REVIEW

**Comment by gigawhitlocks** [comment]
I just want you to know I haven't overlooked this PR, I've just not taken the time to fully digest this change yet and I apologize for the slow review

**Review by gigawhitlocks — APPROVED** [review]
This is really great Szymon, and actually leads me wondering if I'm not fully aware of the pros/cons of doing certain long-running tasks in the Provisioner itself vs leveraging k8s Jobs like you have here, because this looks like a very robust implementation to me.

That said: if you see me doing unnecessary work where I could be taking advantage of what  Kubernetes in my PRs, please let me know :)

File: internal/provisioner/backup_restore.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+	return &BackupOperator{
+		jobTTLSecondsAfterFinish: jobTTL,
+		backupRestoreImage:       image,
+		awsRegion:                region,
+	}
+}
+
+// TriggerBackup creates new backup job and waits for it to start.
+func (o BackupOperator) TriggerBackup(
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply internal/provisioner/backup_restore.go:57]: Good question @gigawhitlocks! Yes, it is intentional although it is not very important, I like this approach as it is more explicit that the state of the receiver - in this case `BackupOperator` - is not modified by this method.

**Review by gabrieljackson — COMMENTED** [review]
Sorry about the bifrost manifest conflict. Request me review when that is addressed and I will review this asap.

File: cmd/cloud/server.go [hunk 2]
```diff
@@ -224,16 +233,21 @@ var serverCmd = &cobra.Command{
 
 		resourceUtil := utils.NewResourceUtil(instanceID, awsClient)
 
+		provisioningParams := provisioner.ProvisioningParams{
```
[the remark this answers is not in the document]
Comment (gabrieljackson) [reply cmd/cloud/server.go:236]: Yep. +1 on nice cleanup.

File: internal/provisioner/kops_provisioner_backup_restore.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	defer invalidateCache(err)
+
+	filestoreCfg, filestoreSecret, err := provisioner.resourceUtil.GetFilestore(installation).
+		GenerateFilestoreSpecAndSecret(provisioner.store, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get files store configuration for installation")
+	}
+	// InstallationBackup is not supported for local MinIO storage, therefore this should not happen
+	if filestoreCfg == nil || filestoreSecret == nil {
+		return nil, errors.New("file store secret and config cannot be empty for backup")
```
Comment (gabrieljackson) [inline internal/provisioner/kops_provisioner_backup_restore.go:35]: nit:
```suggestion
		return nil, errors.New("filestore secret and config cannot be empty for backup")
```

File: internal/provisioner/kops_provisioner_backup_restore.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	defer invalidateCache(err)
+
+	jobsClient := k8sClient.Clientset.BatchV1().Jobs(backup.InstallationID)
+
+	return provisioner.backupOperator.CheckBackupStatus(jobsClient, backup, logger)
+}
+
+// CleanupBackup deletes backup job from the cluster if it exists.
+func (provisioner *KopsProvisioner) CleanupBackup(backup *model.InstallationBackup, cluster *model.Cluster) error {
```
Comment (gabrieljackson) [inline internal/provisioner/kops_provisioner_backup_restore.go:73]: nit: Maybe `CleanupBackupJob` is a bit of a better name for this.

File: internal/provisioner/kops_provisioner_backup_restore.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	logger := provisioner.logger.WithFields(log.Fields{
+		"cluster":      cluster.ID,
+		"installation": backup.InstallationID,
+		"backup":       backup.ID,
+	})
+	logger.Info("Cleaning up backup for installation")
```
Comment (gabrieljackson) [inline internal/provisioner/kops_provisioner_backup_restore.go:79]: nit: same here. It could look like a backup itself is being deleted.

**Review by gabrieljackson — APPROVED** [review]
This is great! No blockers on my end.

File: internal/provisioner/kops_provisioner_backup_restore.go (hunk 4 above)
Comment (Szymongib) [reply internal/provisioner/kops_provisioner_backup_restore.go:73]: I did not want to expose the internal mechanism outside meaning that it runs a job, but I agree that the `CleanupBackup` might be a bit misleading.

**Review by angeloskyratzakos — APPROVED** [review]
LGTM

**Pull request merged**